kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,700 @@
+	} else {
+		// Type is load balancer, this means calico annotations have changed, and we need to re-assign IPs from this service
+		err := c.releaseIP(createHandle(svc))
+		if err != nil {
+			return err
+		}
+		// We can only assign new IP if we are still managing the address assignment
+		if isCalicoManagedLoadBalancer(svc, c.cfg.AssignIPs) {
+			return c.assignIP(svc)
+		} else {
+			// No longer managed by Calico, clean up our assigned IP from Service
+			svc.Status.LoadBalancer.Ingress = []v1.LoadBalancerIngress{}
+
+			_, err = c.clientSet.CoreV1().Services(svc.Namespace).UpdateStatus(c.ctx, svc, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+}
+
+// resolvePools valid IPpool range when specific pool is requested by the user
+func (c *loadBalancerController) resolvePools(pools []string, isv4 bool) ([]cnet.IPNet, error) {
+	// Iterate through the provided pools. If it parses as a CIDR, just use that.
+	// If it does not parse as a CIDR, then attempt to lookup an IP pool with a matching name.
+	result := []cnet.IPNet{}
+	for _, p := range pools {
+		_, cidr, err := net.ParseCIDR(p)
+		if err != nil {
+			// Didn't parse as a CIDR - check if it's the name
+			// of a configured IP pool.
+			for _, ipp := range c.ipPools {
+				if ipp.Name == p {
+					// Found a match. Use the CIDR from the matching pool.
+					_, cidr, err = net.ParseCIDR(ipp.Spec.CIDR)
+					if err != nil {
+						return nil, fmt.Errorf("failed to parse IP pool cidr: %s", err)
+					}
+					log.Infof("Resolved pool name %s to cidr %s", ipp.Name, cidr)
+				}
+			}
+
+			if cidr == nil {
+				// Unable to resolve this pool to a CIDR - return an error.
+				return nil, fmt.Errorf("error parsing pool %q: %s", p, err)
+			}
+		}
+
+		ip := cidr.IP
+		if isv4 && ip.To4() == nil {
+			return nil, fmt.Errorf("%q isn't a IPv4 address", ip)
+		}
+		if !isv4 && ip.To4() != nil {
+			return nil, fmt.Errorf("%q isn't a IPv6 address", ip)
+		}
+		result = append(result, cnet.IPNet{IPNet: *cidr})
+	}
+	return result, nil
+}
+
+// isCalicoManagedLoadBalancer returns if Calico should try to assign IP address for the LoadBalancer
+// We assign IPs only if the loadBalancer controller assignIP is set to AllService
+// or in RequestedOnlyServices if the service has calico annotation
+func isCalicoManagedLoadBalancer(svc *v1.Service, assignIPs string) bool {
+	if svc.Spec.Type == v1.ServiceTypeLoadBalancer {
+		if assignIPs == api.AllServices {
+			return true
+		}
+
+		if assignIPs == api.RequestedServicesOnly {
+			if svc.Annotations[annotationIpv4Pools] != "" ||
+				svc.Annotations[annotationIpv6Pools] != "" ||
+				svc.Annotations[annotationLoadBalancerIp] != "" {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// validateAnnotation checks if the ips specified in the calico annotation are valid.
+// Each service can have at most one ipv4 and one ipv6 address
+func validateAnnotation(ipAddrs []string) ([]cnet.IP, error) {
```

Comment:
This function is called `validateAnnotation` but it's actually validating a list of strings.

I'd expect `validateAnnotation` to take the raw annotation - I'd probably call this `parseCIDRs`